seqfold/fold.go

```diff
@@ -0,0 +1,875 @@
+package seqfold
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"strings"
+)
+
+// Fold the DNA sequence and return the lowest free energy score.
+//
+// Based on the approach described in:
+// Zuker and Stiegler, 1981
+// https://www.ncbi.nlm.nih.gov/pmc/articles/PMC326673/pdf/nar00394-0137.pdf
+//
+// If the sequence is 50 or more bp long, "isolated" matching bp
+// are ignored in V(i,j). This is based on an approach described in:
+// Mathews, Sabina, Zuker and Turner, 1999
+// https://www.ncbi.nlm.nih.gov/pubmed/10329189
+//
+// Args:
+//     seq: The sequence to Fold
+// Keyword Args:
+//     temp: The temperature the Fold takes place in, in Celsius
+// Returns:
+//     List[Struct]: A list of structures. Stacks, bulges, hairpins, etc.
+func Fold(seq string, temp float64) []NucleicAcidStructure {
+	//v_cache, w_cache := cache(seq, temp)
+	//n := len(seq)
+	fc, err := NewFoldingContext(seq, temp)
+	if err != nil {
+		fmt.Printf("%+v\n", err)
+		os.Exit(1)
+	}
+
+	// get the minimum free energy structure out of the cache
+	return Traceback(0, len(seq)-1, fc)
+}
+
+// Fold the sequence and return just the delta G of the structure
+//
+// Args:
+//     seq: The sequence to fold
+// Keyword Args:
+//     temp: The temperature to fold at
+// Returns:
+//     float: The minimum free energy of the folded sequence
+func dG(seq string, temp float64) float64 {
+	structs := Fold(seq, temp)
+	sumE := 0.0
+	for _, s := range structs {
+		sumE += s.E
+	}
+	return RoundFloat(sumE, 2)
+}
+
+// Fold a nucleic acid sequence and return the estimated dg of each (i,j) pairing.
+//
+// Args:
+//     seq: The nucleic acid sequence to fold
+// Keyword Args:
+//     temp: The temperature to fold at
+// Returns:
+//     Cache: A 2D matrix where each (i, j) pairing corresponds to the
+//         minimum free energy between i and j
+func dGCache(seq string, temp float64) ([][]float64, error) {
+	//_, w_cache := cache(seq, temp)
+	fc, err := NewFoldingContext(seq, temp)
+	if err != nil {
+		return nil, fmt.Errorf("error: %w", err)
+	}
+	wCache := fc.W
+
+	cache := make([][]float64, len(seq))
+	for i := range cache {
+		cache[i] = make([]float64, len(seq))
+	}
+
+	for i, wrow := range wCache {
+		for j, s := range wrow {
+			cache[i][j] = s.E
+		}
+	}
+
+	return cache, nil
+}
+
+// Get the dot bracket notation for a secondary structure.
```

Comment:
need to find explainer of dot bracket notation.

Reply:
[This](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4041456/#:~:text=The%20most%20common%20encoding%20for,code%20for%20an%20unpaired%20base%20'.) might help but I did not manage to find a good source that explains the notation.